### PR TITLE
Videos: Use start time and end time for clips

### PR DESCRIPTION
**Pull request by ChunkyProgrammer**
This PR parses the start and end time for clips. 
closes https://github.com/iv-org/invidious/issues/3921
https://github.com/FreeTubeApp/FreeTube/issues/4323

Ex url: [https://www.youtube.com/clip/UgkxxPM3BRphCAPLP88YoUGuj79KXPfpNNO_](https://www.youtube.com/clip/UgkxxPM3BRphCAPLP88YoUGuj79KXPfpNNO_)
API url: `/api/v1/clips/UgkxxPM3BRphCAPLP88YoUGuj79KXPfpNNO_?pretty=1`

API Response:

```typescript
{
  "startTime": 8842.645,
  "endTime": 8855.856,
  "clipTitle": "✂️ Kirby is pink!",
  "video": [Video Object](https://docs.invidious.io/api/common_types/#videoobject)
}
```

### CODE REVIEW

**Review by syeopite — APPROVED** [review]

File: src/invidious/videos/clip.cr [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  start_time = decoded_protobuf
+    .try(&.["50:0:embedded"]["2:1:varint"].as_i64)
+
+  end_time = decoded_protobuf
+    .try(&.["50:0:embedded"]["3:2:varint"].as_i64)
+
+  clip_title = decoded_protobuf
+    .try(&.["50:0:embedded"]["4:3:string"].as_s)
+
+  return (start_time / 1000), (end_time / 1000), clip_title
```
Comment (SamantazFox) [inline src/invidious/videos/clip.cr:19]: I'm surprised you didn't get any compile-time errors here. Your `start_time` and `end_time` variables can be `nil`, so a division on them won't work.

Your return type should be either of those:
 * `{Float64?, Float64?, String?}` if you use `.try` everywhere, including for the divisions (i.e `.try &./ 1000`)
 * `{Float64, Float64, String} | Nil` if you let it fail and rescue the exception at the end of the function

File: src/invidious/routes/watch.cr [hunk 2]
```diff
@@ -275,6 +275,12 @@ module Invidious::Routes::Watch
     return error_template(400, "Invalid clip ID") if response["error"]?
 
     if video_id = response.dig?("endpoint", "watchEndpoint", "videoId")
+      if params = response.dig?("endpoint", "watchEndpoint", "params").try &.as_s
+        start_time, end_time, _ = parse_clip_parameters(params)
+        env.params.query["start"] = start_time.to_s
+        env.params.query["end"] = end_time.to_s
```
Comment (SamantazFox) [inline src/invidious/routes/watch.cr:281]: Depending on how you'll handle an unexpected parsing error at `parse_clip_parameters`, those might need some `nil` checks

File: src/invidious/videos/clip.cr [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  decoded_protobuf = params.try { |i| URI.decode_www_form(i) }
+    .try { |i| Base64.decode(i) }
+    .try { |i| IO::Memory.new(i) }
+    .try { |i| Protodec::Any.parse(i) }
+
+  start_time = decoded_protobuf
+    .try(&.["50:0:embedded"]["2:1:varint"].as_i64)
+    .try { |i| i/1000 }
+
+  # start_time = start_time.to_i64 / 1000 if start_time != nil
```
Comment (SamantazFox) [inline src/invidious/videos/clip.cr:14]: Can you remove those please?

**Review by SamantazFox — APPROVED** [review]

File: src/invidious/routes/api/v1/videos.cr [hunk 4]
```diff
@@ -363,4 +363,47 @@ module Invidious::Routes::API::V1::Videos
...
+    rescue ex : NotFoundException
+      return error_json(404, ex)
+    rescue ex
+      return error_json(500, ex)
+    end
+
+    return JSON.build do |json|
+      json.object do
+        json.field "startTime", start_time
+        json.field "endTime", end_time
```
Comment (SamantazFox) [inline src/invidious/routes/api/v1/videos.cr:401]: Small change, to make the field names consistent with #4205 and youtube's side in general:
```suggestion
        json.field "startTimeSeconds", start_time
        json.field "endTimeSeconds", end_time
```

**Comment by SamantazFox** [comment]
For some reason, Github didn't automatically close this PR when it was merged...

**Pull request closed without merging**